File: haskellnotebook/haskell/HaskellSimpleTextMining/tools/misc/dbreader/scala/src/scala/dbreader.scala
//************************************************
//* Copyright (c) 2007 (botnode.com).  All Rights Reserved
//* 
//* Created On: 11/6/2007
//* 
//* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
//* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
//* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
//* A PARTICULAR PURPOSE ARE DISCLAIMED.
//* 
//* (see /LICENSE for more details)
//************************************************
/*
 data SpiderDatabase =  SpiderDatabase { 
      magicNumberA :: Word16,
      magicNumberB :: Word16,
      majorVers :: Word16,
      minorVers :: Word16,
      headerTag :: Word16,
      poolLen :: Word16,
      spiderpool :: [URLSet]
    }
data URLSet = URLSet {
      urlinfo :: URLInfo,
      titleinfo :: TitleInfo,
      descrinfo :: DescrInfo,
      keywordsinfo :: KeywordsInfo
}
data URLInfo = URLInfo {
      tag :: Word8,
      urlid :: Word16,
      urllen :: Word16,
      url :: ByteString
}
data TitleInfo = TitleInfo {
      titletag :: Word8,      
      titlelen :: Word16,
      title :: ByteString
}
data DescrInfo = DescrInfo {
      descrtag :: Word8,      
      descrlen :: Word16,
      descr :: ByteString
}
data KeywordsInfo = KeywordsInfo {
      keywordstag :: Word8,      
      keywordslen :: Word16,
      keywords :: ByteString
}
*/

package org.botlist.dbreader

import scala.Console
import scala.io.Source
import java.io._

object DbReader {
  
  def main(args: Array[String]): Unit = {
	Console.println("INFO: Indexing Document Data <standby> ...")
	val timeStart = System.currentTimeMillis()	   

	val fis = new FileInputStream( args(0) )
	val bis = new BufferedInputStream( fis, 4096 )
    val inputstream = new DataInputStream( bis )
	
	val magicNumberA = inputstream.readUnsignedShort()
	val magicNumberB = inputstream.readUnsignedShort()	
	val versMajor = inputstream.readUnsignedShort()
	val versMinor = inputstream.readUnsignedShort()
	val headerTag = inputstream.readUnsignedShort()
	val poolct = inputstream.readUnsignedShort()
    Console.printf("Magic Number A: 0x{0} 0x{1}\n", 
				   Integer.toHexString(magicNumberA), 
				   Integer.toHexString(magicNumberB)) 
    Console.printf("Version: 0x{0} 0x{1}\n", 
				   Integer.toHexString(versMajor), 
				   Integer.toHexString(versMinor))	
	Console.printf("Header Tag: 0x{0}\n", 
				   Integer.toHexString(headerTag))
	Console.printf("Pool Len Count: {0,number}\n", poolct)
	
	// Extract the URL content, first URL
	val tag = inputstream.readUnsignedByte()
	val urlid = inputstream.readUnsignedShort()
	val urllen = inputstream.readUnsignedShort()

	Console.printf("URL Info: 0x{0}, {1}, {2}\n", 
				   Integer.toHexString(tag),
				   urlid, urllen)
	val bytestrdata = new Array[byte](urllen)
	inputstream.readFully(bytestrdata)
	val urlstr = new String(bytestrdata)
	Console.printf("URL: [{0}]\n", urlstr)

	// Read the title info struct
	val titletag = inputstream.readUnsignedByte()
	val titlelen = inputstream.readUnsignedShort()
	Console.printf("Title Info: 0x{0}, {1}\n",
				   Integer.toHexString(titletag), titlelen)
	
	// Print the title string data
	val bytestrdata2 = new Array[byte](titlelen)
	inputstream.readFully(bytestrdata2)
	val titlestr = new String(bytestrdata2)
	Console.printf("Title: [{0}]\n", titlestr)
	
	inputstream.close()
    val timeEnd = System.currentTimeMillis()
    Console.println("Done...")
    Console.println("Completed processing in " + (timeEnd - timeStart) + " ms.")
  }
}

File: haskellnotebook/haskell/HaskellSimpleTextMining/tools/misc/dbreader/haskell/DbReader.hs
--
-- Author: Berlin Brown
-- File: DbReader.hs
--
-- Also see: 
-- (1) http://www.zvon.org/other/haskell/Outputio/index.html
-- (2) http://hackage.haskell.org/packages/archive/binary/0.4.1/doc/html/Data-Binary.html
-- (3) http://blog.kfish.org/2007/10/survey-haskell-unicode-support.html
-- (4) http://hackage.haskell.org/cgi-bin/hackage-scripts/package/utf8-string-0.2
-- (5) http://hackage.haskell.org/packages/archive/bytestring/0.9.0.1/doc/html/Data-ByteString.html
-- Useful endian loading functions:
-- getWord8, getWord16be, getWord32be
--

module Main where

import Data.Word
import Data.Binary
import qualified Data.ByteString.Lazy.Char8 as LazyChar8 (unpack)
import Data.ByteString.Lazy as Lazy (ByteString, unpack)
import Data.ByteString (unpack)
--import Codec.Binary.UTF8.String as Unicode
import Data.Binary.Get as BinaryGet
import Data.Binary.Put as BinaryPut
import IO
import Text.Printf
import System.Environment
import Control.Monad (replicateM, liftM)

{- *********************************************************
     Define the Database Data Types
     SpiderDatabase represents a singleton wrapper for an
     entire database.
   ********************************************************* -}
data SpiderDatabase =  SpiderDatabase { 
      magicNumberA :: Word16,
      magicNumberB :: Word16,
      majorVers :: Word16,
      minorVers :: Word16,
      headerTag :: Word16,
      poolLen :: Word16,
      spiderpool :: [URLSet]
    }
data URLSet = URLSet {
      urlinfo :: URLInfo,
      titleinfo :: TitleInfo,
      descrinfo :: DescrInfo,
      keywordsinfo :: KeywordsInfo
}
data URLInfo = URLInfo {
      tag :: Word8,
      urlid :: Word16,
      urllen :: Word16,
      url :: ByteString
}
data TitleInfo = TitleInfo {
      titletag :: Word8,      
      titlelen :: Word16,
      title :: ByteString
}
data DescrInfo = DescrInfo {
      descrtag :: Word8,      
      descrlen :: Word16,
      descr :: ByteString
}
data KeywordsInfo = KeywordsInfo {
      keywordstag :: Word8,      
      keywordslen :: Word16,
      keywords :: ByteString
}
{- *********************************************************
     Class instances
   ********************************************************* -}
instance Show SpiderDatabase where
    show db = let magicb = (magicNumberB db)
                  header = (headerTag db)
                  poolct = (poolLen db)
              in "<<<Database Content>>>\n" ++
                 (((printf "Magic: %X %X\n") (magicNumberA db)) (magicNumberB db)) ++
                 printf "URL Pool Count: %d\n" poolct ++
                 printf "Header Tag: %X\n" header ++
                 printf "URL Tag %X\n" a ++
                 printf "URL Idx %X\n" b ++
                 printf "URL: %s\n" c ++
                 printf "Title: %s\n" e ++
                 "<<<End>>>"
              where x = (spiderpool db)
                    y = (x !! 0)
                    z = (urlinfo y)
                    a = (tag z)
                    b = (urlid z)
                    c = (LazyChar8.unpack (url z))
                    d = (titleinfo y)
                    e = (LazyChar8.unpack (title d))

instance Binary URLInfo where
    put _ = do BinaryPut.putWord8 0
    get = do
      urltag <- getWord8
      idx <- getWord16be
      len <- getWord16be
      strdata <- BinaryGet.getLazyByteString (fromIntegral len)
      return (URLInfo {tag=urltag, urlid=idx, 
                       urllen=len, url=strdata})
instance Binary DescrInfo where
    put _ = do BinaryPut.putWord8 0
    get = do
      tag <- getWord8
      len <- getWord16be
      strdata <- BinaryGet.getLazyByteString (fromIntegral len)
      return (DescrInfo {descrtag=tag,
                         descrlen=len, 
                         descr=strdata})
instance Binary TitleInfo where
    put _ = do BinaryPut.putWord8 0
    get = do
      tag <- getWord8
      len <- getWord16be
      strdata <- BinaryGet.getLazyByteString (fromIntegral len)
      return (TitleInfo {titletag=tag,
                         titlelen=len, 
                         title=strdata})
instance Binary KeywordsInfo where
    put _ = do BinaryPut.putWord8 0
    get = do
      tag <- getWord8
      len <- getWord16be
      strdata <- BinaryGet.getLazyByteString (fromIntegral len)
      return (KeywordsInfo {keywordstag=tag,
                            keywordslen=len, 
                            keywords=strdata})

instance Binary URLSet where
    put _ = do BinaryPut.putWord8 0
    get = do
      i :: URLInfo <- get :: Get URLInfo
      j :: TitleInfo <- get :: Get TitleInfo
      k :: DescrInfo <- get :: Get DescrInfo
      x :: KeywordsInfo <- get :: Get KeywordsInfo
      return (URLSet {urlinfo=i, titleinfo=j, 
                      descrinfo=k, keywordsinfo=x})

getURLSet :: Get URLSet
getURLSet = get :: Get URLSet

instance Binary SpiderDatabase where
    put _ = do BinaryPut.putWord8 0
    get = do 
      magicnumbera <- BinaryGet.getWord16be
      magicnumberb <- BinaryGet.getWord16be
      major <- BinaryGet.getWord16be
      minor <- BinaryGet.getWord16be
      header <- BinaryGet.getWord16be
      poolct <- BinaryGet.getWord16be
      -- *******************************
      -- Get the remaining byte string data,
      -- So that we can use lazy bytestring to load to load the
      -- the data types.
      -- *******************************
      pool1 <- getURLSet
      return (SpiderDatabase {magicNumberA=magicnumbera,
                              magicNumberB=magicnumberb,
                              majorVers=major,
                              minorVers=minor,
                              headerTag=header,
                              poolLen=poolct,
                              spiderpool=(pool1 : [])
                             })
main = do
  putStrLn "Running Spider Database Reader"
  args <- getArgs
  db :: SpiderDatabase  <- decodeFile (args !! 0)  
  putStrLn $ show db
  putStrLn "Done"
